### ISSUE TITLE
Some initial feedback

### ISSUE THREAD

**Issue by meevans1** [manthony-42/ATLAS_UK_NN_TUTORIAL#1]
These are all just my opinion so please only implement those with which you agree!
Happy to discuss any of them further.
Some points have since been pull requested, marked as SOLVED

1. SOLVED. Check "After doing this,italicised text run"
2. In the cell `stop_file = `, consider deleting the last line and make it an exercise for the user like "do the same for background and call it top_tree". I understand if you don't wanna do this to get straight to exercises on ML and save time.
3. SOLVED. You can remove 'scaleFactor_PHOTON', 'scaleFactor_TAU', 'scaleFactor_BTAG', from `weights` since photons, taus and btagging aren't used. Also simplifies things!
4. SOLVED. Consider changing your linspace commands to (xlow, xhigh, num=bins+1). I try to write the function parameter when I can, to make the function call more transparent. (Also in `do_predictions`)
4. SOLVED. I think you can delete the line `plt.grid(False)` as it doesn't do anything.
5. SOLVED. I think you can change to `plt.legend()` since `loc='best'` in the default anyway.
6. SOLVED. I think you can delete the log functionality from `make_plot` since you never use it.
4. Consider deleting some `make_plot` commands and give exercises like "plot dphi_ll_met from 0 to 4 with 8 bins". I understand if you don't wanna do this to get straight to exercises on ML and save time.
5. SOLVED. You say "samples of events for training and validation" but you call your variables `X_test`. Consider making the name consistent, like `X_val`.
5. Consider deleting layer_3 and output_layer and give exercises "add another layer with 10 nodes, relu activation and call it layer_3", "add another layer with 1 node, sigmoid activation and call it output_layer"
6. SOLVED. You say "There are 3 hidden layers", but you've already talked about the input layer. I think this needs a bit more detail to avoid being confusing. e.g. list which layers are the hidden ones.
7. I like that you've masked the cells `NN_learning_rate` and `Details of the activation function`. (If I've understood correctly) Consider making it clearer that these are optional/extension parts. e.g. write in text "These sections NN learning rate and Details of the activation function are intended as extensions, come back to them later once you've been through the whole notebook once. So skip these sections for now and come back later.". I would in fact prefer moving these sections to just above where you set those exercises, because when I click through the cells with Shift+Enter, these sections are run. Whichever solution you prefer.
8. SOLVED. In `doValidationPlots` is there a reason you make the plots side-by-side? I think underneath each other might be easier as the x-axes are the same, so one can compare easier by imagining a vertical line. This will also simplify the code because you won't need to use subplots, you can just make 2 separate plots.
9. SOLVED. Add `.copy()` at the end of `sig_df` and `bkg_df` to avoid the pandas warnings.
10. SOLVED. Consider adding `plt.legend()` in your cell starting with `lumi = `
11. SOLVED. Consider adding lines:
'''
from tensorflow.random import set_seed # import set_seed function for TensorFlow
seed_value=42
set_seed(seed_value) # set TensorFlow random seed
'''
before any TensorFlow commands for better reproducibility.
8. SOLVED. Consider removing the cell outputs before the tutorial so that students don't see the "answers" when they open the notebook.

I haven't been through the exercises yet in detail so I may have more comments when I get there.

**Comment by meevans1** [comment]
Some further points:
(All points solved in pull request)
1. SOLVED. Perhaps you can get rid of the `pydot` and `graphviz` pip installs and imports? I can't tell where you use them.
11. SOLVED. Consider removing the `color=` commands in your `plt.plot(history_` commands to make the colors and calls consistent within the cell and with the next cell.
12. SOLVED. Clarify your line `Change the activation functions in the hidden layer to`. Don't you have multiple hidden layers?
13. SOLVED. Consider making a single `model` function with `activation_function` and `learning_rate` as callable parameters rather than separate functions like `NNmodel_sigmoid`. This would save multiple lines of code

**Comment by meevans1** [comment]
Some further points:
(Some points solved in pull request)
1. SOLVED. I think you can get rid of `from sklearn import datasets` since you don't use it
2. SOLVED. I think you can get rid of `to_categorical` in `from keras.utils import` since you don't use it
3. SOLVED. I think you can get rid of `ModelCheckpoint` in `from keras.callbacks import` since you don't use it
4. SOLVED. I think you can get rid of `Activation` in `from keras.layers import` since you don't use it
5. SOLVED. I think you can get rid of pip install and import `awkward1` since you don't use it
6.SOLVED.  I think you can get rid of `pip install grpcio` since you don't use it
7. SOLVED. I think you can remove pip install keras and sklearn since they're already in this default version of colab.
6. SOLVED. Are `progressbar2` and `progressbar` the same or different? Because you pip install one and import the other. Be consistent. progressbar2 and progressbar are both already in the default version of colab anyway, so you don't need to pip install
7. SOLVED. Consider switching the flag `first_run` to False after you've downloaded the input files.
8. SOLVED. Check "Activation functions For each neuron". Some typo?
9. In the Activation functions section, consider defining the variables like "where w is weights..."
10.SOLVED.  In the Activation functions section, consider check "non-reducible structure, a A more useful". Some typo?
11. SOLVED. You could remove the `color` parameters in `plt.plot(range_array` since matplotlib will automatically assign colours
12. Consider moving the whole "Details of the activation function" right to the end after the exercises.
13. Consider making the "Details of the activation function" into more exercises. e.g. Using the equations in the text, define functions for sigmoid, ReLU, LeakyReLU... 
14. SOLVED. You can remove `import matplotlib` from PlotActivations

**Comment by meevans1** [comment]
Points 3,5,6,7 of first comment solved in pull request https://github.com/manthony-42/ATLAS_UK_NN_TUTORIAL/pull/5

**Comment by meevans1** [comment]
Points 1,4,9,14,15,16 from first comment, 1,2,3,4 from second comment, 1,2,3,4,5,6,7,8,9,10,13,16 from third comment implemented in pull request https://github.com/manthony-42/ATLAS_UK_NN_TUTORIAL/pull/6

**Comment by meevans1** [comment]
Points 11,13 from first comment implemented in pull request #6
